nomad/host_volume_endpoint_test.go

```diff
@@ -763,6 +764,113 @@ func TestHostVolumeEndpoint_placeVolume(t *testing.T) {
+
+	c, node := newMockHostVolumeClient(t, srv, "default")
+
+	vol := &structs.HostVolume{
+		Name:      "test-vol",
+		Namespace: "default",
+		NodeID:    node.ID,
+		PluginID:  "mkdir",
+		HostPath:  "/pretend/path",
+		RequestedCapabilities: []*structs.HostVolumeCapability{
+			{
+				AttachmentMode: structs.HostVolumeAttachmentModeFilesystem,
+				AccessMode:     structs.HostVolumeAccessModeSingleNodeWriter,
+			},
+		},
+	}
+	wr := structs.WriteRequest{Region: srv.Region()}
+
+	// create the volume for us to attempt concurrent operations on
+	c.setCreate(&cstructs.ClientHostVolumeCreateResponse{
+		VolumeName: "test-vol",
+		HostPath:   "/pretend/path",
+	}, nil)
+	createReq := &structs.HostVolumeCreateRequest{
+		Volume:       vol.Copy(), // copy because HostPath gets mutated
+		WriteRequest: wr,
+	}
+	var createResp structs.HostVolumeCreateResponse
+	must.NoError(t, srv.RPC("HostVolume.Create", createReq, &createResp))
+	volumeID := createResp.Volume.ID // used by delete
+
+	// prepare blocking channels in the mock client.
+	// sending a struct{} (or closing) the channel will unblock the operation
+	cCh, rCh, dCh := make(chan struct{}), make(chan struct{}), make(chan struct{})
+	c.setBlockChan("create", cCh)
+	c.setBlockChan("register", rCh)
+	c.setBlockChan("delete", dCh)
+
+	// each RPC call that finishes will put its name here
+	opCh := make(chan string)
+
+	// start all the RPCs concurrently
+	go func() {
+		createReq = &structs.HostVolumeCreateRequest{
+			Volume:       vol.Copy(), // copy because HostPath gets mutated
+			WriteRequest: wr,
+		}
+		createResp = structs.HostVolumeCreateResponse{}
+		test.NoError(t, srv.RPC("HostVolume.Create", createReq, &createResp),
+			test.Sprint("Create method should not error"))
+		opCh <- "create"
+	}()
+	go func() {
+		registerReq := &structs.HostVolumeRegisterRequest{
+			Volume:       vol.Copy(),
+			WriteRequest: wr,
+		}
+		var registerResp structs.HostVolumeRegisterResponse
+		test.NoError(t, srv.RPC("HostVolume.Register", registerReq, &registerResp),
+			test.Sprint("Register method should not error"))
+		opCh <- "register"
+	}()
+	go func() {
+		deleteReq := &structs.HostVolumeDeleteRequest{
+			VolumeID:     volumeID,
+			WriteRequest: wr,
+		}
+		var deleteResp structs.HostVolumeDeleteResponse
+		test.NoError(t, srv.RPC("HostVolume.Delete", deleteReq, &deleteResp),
+			test.Sprint("Delete method should not error"))
+		opCh <- "delete"
+	}()
+
+	// helper pulls an operation from the channel, or timeout
+	pullOp := func() string {
+		select {
+		case op := <-opCh:
+			return op
+		case <-time.After(200 * time.Millisecond): // generous headroom
+			return "timeout"
+		}
+	}
+
+	must.Eq(t, "timeout", pullOp(), must.Sprint("nothing should be unblocked yet"))
+
+	close(rCh)
+	must.Eq(t, "register", pullOp(), must.Sprint("closing register channel should unblock Register"))
+
+	must.Eq(t, "timeout", pullOp(), must.Sprint("again blocked RPCs should remain so"))
```

Comment:
At first look I thought this timeout was effectively testing the test mock's blocking and not the RPC blocking behavior. But really this prevents the test from getting stuck in case of a failure of one of the goroutines. So :+1: 

Because we know the following client RPCs are blocked, I'm pretty sure we can instead query the state store to verify the raft write has completed here too if we want? Ex. at this line we know the volume has been registered but not created (maybe we could set `parameters` on the volume object in the request to have something to check quickly).

Reply:
ah sure, I'll add a bit to check some unique state based on the operation

Reply:
tangent here: is there a reason we exclude `Parameters` from `HostVolumeStub` that gets returned by the `List` RPC?
https://github.com/hashicorp/nomad/blob/46bd0b171664641144ac5bcc59dbd8e8c1f45c8c/nomad/structs/host_volumes.go#L352-L368

I don't need it for this test, but I bumped into it on the way down a misguided road.

Reply:
Stubs are minimal representations -- typically only what you minimally need for the CLI/UI list view (and handling blocking queries). Anything like capabilities, parameters, constraints, etc. are good candidates to omit here.

Reply:
Ah right `v.volOps` doesn't let us know which. This is a little rough... we can't force the request goroutines to make their calls in order with `sync.Wait`, because we don't return from the server RPC call till we're done. But we _do_ know when a given RPC handler has started the client RPC, because we can read `mockHostVolumeClient.blockChans`. So with that we could do this:

* Start the register goroutine
* Wait until `mockHostVolumeClient.blockChans` is blocked on `register`
* Start the create goroutine (which we know is blocked)
* Start the delete goroutine (which we know is blocked)
* Unblock register
* Wait until `mockHostVolumeClient.blockChans` is blocked on _either_ create or delete
* Unblock whichever channel is blocked